src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1043,139 +1068,637 @@ private StringConcatFactory() {
+         *      int arg0, long arg1, boolean arg2, char arg3, String arg5)
+         * </pre></blockquote>
+         */
+        private static MethodTypeDesc prependArgs(MethodType concatArgs) {
+            int parameterCount = concatArgs.parameterCount();
+            var paramTypes = new ClassDesc[parameterCount + 4];
+            paramTypes[0] = CD_int;          // length
+            paramTypes[1] = CD_byte;         // coder
+            paramTypes[2] = CD_Array_byte;   // buff
+            paramTypes[3] = CD_Array_String; // constants
+
+            for (int i = 0; i < parameterCount; i++) {
+                var cl = concatArgs.parameterType(i);
+                paramTypes[i + 4] = needStringOf(cl) ? CD_String : ConstantUtils.classDesc(cl);
+            }
+            return MethodTypeDesc.of(CD_int, paramTypes);
+        }
 
-        private SimpleStringBuilderStrategy() {
-            // no instantiation
+        /**
+         * Construct the MethodType of the coder method,
+         * The first parameter is the initialized coder, Only parameter types that can be UTF16 are added.
+         */
+        private static MethodTypeDesc coderArgs(MethodType concatArgs) {
+            int parameterCount = concatArgs.parameterCount();
+            List<ClassDesc> paramTypes = new ArrayList<>();
+            paramTypes.add(CD_int); // init coder
+            for (int i = 0; i < parameterCount; i++) {
+                var cl = concatArgs.parameterType(i);
+                if (maybeUTF16(cl)) {
+                    paramTypes.add(cl == char.class ? CD_char : CD_String);
+                }
+            }
+            return MethodTypeDesc.of(CD_int, paramTypes);
+        }
+
+        /**
+         * Construct the MethodType of the length method,
+         * The first parameter is the initialized length
+         */
+        private static MethodTypeDesc lengthArgs(MethodType concatArgs) {
+            int parameterCount = concatArgs.parameterCount();
+            var paramTypes = new ClassDesc[parameterCount + 1];
+            paramTypes[0] = CD_int; // init long
+            for (int i = 0; i < parameterCount; i++) {
+                var cl = concatArgs.parameterType(i);
+                paramTypes[i + 1] = needStringOf(cl) ? CD_String : ConstantUtils.classDesc(cl);
+            }
+            return MethodTypeDesc.of(CD_int, paramTypes);
         }
 
         private static MethodHandle generate(Lookup lookup, MethodType args, String[] constants) throws Exception {
-            String className = getClassName(lookup.lookupClass());
+            lookup = new MethodHandles.Lookup(String.class);
+            String className = "java.lang.String$$StringConcat";
+            final MethodType concatArgs = erasedArgs(args);
+
+            // 1 argment use built-in method
+            int paramCount  = args.parameterCount();
+            var concatClass = ConstantUtils.binaryNameToDesc(className);
```

Comment:
`String className` and `ClassDesc concatClass` are effectively constant. Refactor to `static final` to save a few instructions.